### PR TITLE
feat(async_vllm): add async vllm worker and unified sync/async interfaces

### PR DESCRIPTION
**Pull request by anHappyDog**
### Description
Currently RLinf only support sync vLLm worker, add async vLLM worker and unify sync/async interfaces

### How has this been tested?

Test 1.5B math model async/sync and 7B math model sync in single machine.

Below is 8 machine 's reward where vllm pipeline's reward is corresponding to sglang' pipeline's reward.

img_v3_02qk_1080c56e-da47-4b76-b0f9-f651b182ef9g

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: rlinf/workers/rollout/vllm/vllm_worker.py [hunk 1]
```diff
@@ -12,25 +12,36 @@
...
 from rlinf.utils.placement import ComponentPlacement
 from rlinf.workers.rollout.utils import print_vllm_outputs
+from toolkits.math_verifier.verify import MathRewardModel, math_verify_call
 
-from . import VLLMEngine
+from . import VLLMExecutor
 
 
-class VLLMWorker(Worker):
+class AsyncVLLMWorker(Worker):
```
Comment (andylin-hao) [inline rlinf/workers/rollout/vllm/vllm_worker.py]: No need to call it `Async`, simply `VLLMWorker` will do

File: rlinf/workers/rollout/vllm/vllm_worker.py [hunk 2]
```diff
@@ -93,39 +108,168 @@ def _get_sampling_params_from_config(self) -> SamplingParams:
                 repetition_penalty=cfg_sampling_params.repetition_penalty,
                 max_tokens=cfg_sampling_params.max_new_tokens,
                 output_kind=RequestOutputKind.FINAL_ONLY,
+                n=self._cfg.algorithm.group_size,
+                logprobs=0 if self._return_logprobs else None,
             )
         return sampling_params
 
-    def _validate_weight_at_first(self) -> None:
+    def _process_image_data(
```
Comment (andylin-hao) [inline rlinf/workers/rollout/vllm/vllm_worker.py:122]: Can this be used for SGLang as well? If so, make this a utility function or class; otherwise it's fine

File: rlinf/workers/rollout/vllm/vllm_worker.py [hunk 3]
```diff
@@ -93,39 +108,168 @@ def _get_sampling_params_from_config(self) -> SamplingParams:
...
         """
+        generation_tasks = []
         if self._cfg.rollout.detokenize:
-            outputs = self._vllm_engine.generate(
-                input_ids=None,
-                sampling_params=self._validate_sampling_params,
-                return_logprobs=False,
-                prompt_texts=self._validate_prompts,
-            )
+            for prompt in self._validate_prompts:
```
Comment (andylin-hao) [inline rlinf/workers/rollout/vllm/vllm_worker.py]: Merge these two `generate` call. Simply differentiate the parameters:
```
if self._cfg.rollout.detokenize:
    input_ids = None
    prompt_text = prompt
else:
    input_ids = prompt_ids
    prompt_text = None
```

File: rlinf/workers/rollout/vllm/vllm_worker.py (hunk 2 above)
Comment (anHappyDog) [reply rlinf/workers/rollout/vllm/vllm_worker.py:122]: SGLang will do such job inner its engine, while vLLM require `Image object` or `torch.Tensor` in `TokensPrompt`/`TextPrompt` as multimodal data, so currently putting this function into VLLMWorker is ok.

File: rlinf/workers/rollout/vllm/vllm_worker.py [hunk 4]
```diff
@@ -143,72 +308,105 @@ def init_worker(self) -> None:
...
+        rewards_tensor = torch.tensor(rewards, dtype=torch.float)
+        rollout_result.rewards = rewards_tensor.reshape(-1, 1)
 
-        requests: List[RolloutRequest] = request.repeat_and_split(
-            self._rollout_batch_size
+        mean = rewards_tensor.mean()
+        std = rewards_tensor.std(unbiased=False)
+        rollout_result.advantages = (rewards_tensor - mean) / (std + 1e-6)
+
+    async def _do_rollout(self, request: RolloutRequest, output_channel: Channel):
```
Comment (Lin-xs) [inline rlinf/workers/rollout/vllm/vllm_worker.py]: `rollout_and_return`

Comment (anHappyDog) [reply rlinf/workers/rollout/vllm/vllm_worker.py]: done.

File: rlinf/workers/rollout/vllm/vllm_worker.py [hunk 5]
```diff
@@ -93,39 +114,194 @@ def _get_sampling_params_from_config(self) -> SamplingParams:
...
-        print("===============================", flush=True)
+        for request_output in vllm_outputs:
+            print_vllm_outputs(request_output, self._tokenizer)
+
+    async def offload_model_weights(self) -> None:
+        """
+        Use async_engine to offload model weights/kv cache.
+        """
+        await self._async_engine.collective_rpc("offload_model_weights")
+        await self._async_engine.reset_prefix_cache()
```
Comment (Lin-xs) [inline rlinf/workers/rollout/vllm/vllm_worker.py:177]: `reset_prefix_cache` should be called after `sync_model_from_actor` rather than `offload_model_weights` because the latter will not be invoked in pipeline mode.

File: tests/e2e_tests/math/vllm/qwen2.5-1.5b-grpo-pipeline-rollout-logprobs.yaml [hunk 6]
```diff
@@ -0,0 +1,268 @@
...
+  - override hydra/job_logging: stdout
+
+hydra:
+  run:
+    dir: .
+  output_subdir: null
+
+cluster:
+  num_nodes: 1
+  num_gpus_per_node: 8
```
Comment (Lin-xs) [inline tests/e2e_tests/math/vllm/qwen2.5-1.5b-grpo-pipeline-rollout-logprobs.yaml:11]: remove `num_gpus_per_node`

File: rlinf/workers/rollout/vllm/vllm_worker.py [hunk 7]
```diff
@@ -93,39 +114,195 @@ def _get_sampling_params_from_config(self) -> SamplingParams:
...
+        )
+
+        return outputs
+
+    async def init_worker(self) -> None:
+        """
+        Use EngineArgs and VllmConfig to initialize VLLM async engine.
         Then offload the model weights, ready to use weights sent from actor.
         """
         engine_args: EngineArgs = EngineArgs(
```
Comment (Lin-xs) [inline rlinf/workers/rollout/vllm/vllm_worker.py:308]: Add `load_format` to avoid redundant weight loading.

File: rlinf/workers/rollout/utils.py [hunk 8]
```diff
@@ -553,9 +553,7 @@ def get_rollout_backend_worker(
         if placement.placement_mode == PlacementMode.COLLOCATED:
```
Comment (Lin-xs) [inline rlinf/workers/rollout/utils.py]: merge two if-stat

**Review by Lin-xs — APPROVED** [review]
